hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java

```diff
@@ -1369,12 +1369,69 @@ public S3AEncryptionMethods getS3EncryptionAlgorithm() {
   File createTmpFileForWrite(String pathStr, long size,
       Configuration conf) throws IOException {
     initLocalDirAllocatorIfNotInitialized(conf);
-    Path path = directoryAllocator.getLocalPathForWrite(pathStr,
-        size, conf);
+    Path path = directoryAllocator.getLocalPathForWrite(pathStr, size, conf);
     File dir = new File(path.getParent().toUri().getPath());
-    String prefix = path.getName();
-    // create a temp file on this directory
-    return File.createTempFile(prefix, null, dir);
+    return safeCreateTempFile(pathStr, null, dir);
+  }
+
+  // TODO remove this method when hadoop upgrades to a newer version of java than 1.8
+  /**
+   * Ensure that the temp file prefix and suffix don't exceed the maximum number of characters
+   * allowed by the underlying file system. This validation isn't required in Java 9+ since
+   * {@link java.io.File#createTempFile(String, String, File)} automatically truncates file names.
+   *
+   * @param prefix prefix for the temporary file
+   * @param suffix suffix for the temporary file
+   * @param dir directory to create the temporary file in
+   * @return a unique temporary file
+   * @throws IOException
+   */
+  static File safeCreateTempFile(String prefix, String suffix, File dir) throws IOException
+  {
+    // avoid validating multiple times.
+    // if the jvm running is version 9+ then defer to java.io.File validation implementation
+    if(Float.parseFloat(System.getProperty("java.class.version")) >= 53) {
```

Comment:
this should go in org.apache.hadoop.util.Shell; there's already something similar. will need a test somehow.